File: RB101/small_problems/advanced_1/1.rb
#Three ways to do it. With gsub first:

ADJECTIVE = %w(quick lazy sleepy ugly)
NOUN = %w(fox dog head leg)
VERB = %w(jumps lifts bites licks)
ADVERB = %w(easily lazily noisily excitedly)

def madlib(str)
  str.gsub (/%{(.*?)}/) { Kernel.const_get($1.upcase).sample }
end

# Also with delete and const_get

ADJECTIVE = %w(quick lazy sleepy ugly)
NOUN = %w(fox dog head leg)
VERB= %w(jumps lifts bites licks)
ADVERB = %w(easily lazily noisily excitedly)

def madlib(str)
   str.delete("%{}").split.map  do |wrd|
     Kernel.const_get(wrd.upcase).sample
     rescue NameError; wrd
    end.join(" ")
end

# Same as above but with instance_variable_get

@adjective = %w(quick lazy sleepy ugly)
@noun = %w(fox dog head leg)
@verb = %w(jumps lifts bites licks)
@adverb = %w(easily lazily noisily excitedly)

def madlib(str)
   str.delete("{}%").split.map  do |wrd|
     instance_variable_get("@#{wrd}").sample
     rescue NameError; wrd
    end.join(" ")
end

File: JS210/small_problems/advanced1/1.js
const template1 = 'The ${adjective} brown ${noun} ${adverb} ' +
                  '${verb} the ${adjective} yellow ' +
                  '${noun}, who ${adverb} ${verb} his ' +
                  '${noun} and looks around.';

const template2 = "The ${noun} ${verb} the ${noun}'s ${noun}.";

function madlibs(template) {
  const REPLACEMENT_TEXTS = {
  adjective: ['quick', 'lazy', 'sleepy', 'noisy', 'hungry'],
  noun: ['fox', 'dog', 'head', 'leg', 'tail'],
  verb: ['jumps', 'lifts', 'bites', 'licks', 'pats'],
  adverb: ['easily', 'lazily', 'noisily', 'excitedly'],
  };

function replaceText(match) {
  const key = match.replace(/[^a-z]/g, '');
  const index = Math.floor(Math.random() * REPLACEMENT_TEXTS[key].length);
  return REPLACEMENT_TEXTS[key][index];
  }

  return template.replace(/\${[a-z]+}/g, replaceText);
}

console.log(madlibs(template1));
console.log(madlibs(template2));